# V-pipe 

V-pipe is a bioinformatics pipeline that integrates various computational tools for the analysis of viral high-throughput sequencing data. It supports the reproducible analysis of genomic diversity in intra-host virus populations, which is often involved in viral pathogenesis and virulence.

## Getting started

::::{grid} 1 1 1 3
:class-container: text-center
:gutter: 3

:::{grid-item-card}
:link: tutorials/tutorial_0_install
:link-type: doc
:class-header: bg-light

Installation 
^^^

Install V-pipe with conda or docker and get started with the basics.
:::

:::{grid-item-card}
:link: tutorials/tutorial_hiv
:link-type: doc
:class-header: bg-light

HIV tutorial
^^^

Tutorial on analyzing HIV data with V-pipe.

:::

:::{grid-item-card}
:link: tutorials/tutorial_sarscov2
:link-type: doc
:class-header: bg-light

SARS-CoV-2 tutorial
^^^

Tutorial on analyzing SARS-CoV-2 data with V-pipe.
:::


::::

## Support

Are you running into issues? Or want to stay up-to-date with the latest developments? Here are some ways to get in touch:

- Write an [issue](https://github.com/cbg-ethz/V-pipe/issues) on GitHub 
- Ask a question on [Gitter](https://gitter.im/V-pipe/community)
- Subscribe to the [mailing list](https://sympa.ethz.ch/sympa/info/v-pipe-users)
- Directly contact the developers: [v-pipe@bsse.ethz.ch](mailto:v-pipe@bsse.ethz.ch)

## Citing V-pipe

For citing the latest version of V-pipe refer to {cite}`fuhrmann_v-pipe_2023` or the first paper describing the pipeline {cite}`posada-cespedes_v-pipe_2021`.

```{bibliography}
```